### PR TITLE
Add app package upload session table

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:17111)](https://staging.teamyapp.com/teams/107/tasks/17111)

### CODE REVIEW

**Comment by duanxianbo** [comment]
@magicoder10 I did not upload changed wire file

File: core/dao/app_package_upload_session.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type AppPackageUploadSession interface {
+	FindAppPackageUploadSessionWithTx(
+		ct context.Context,
+		tx *transaction.Transaction,
+		appID uint64,
+		userID uint64,
+		versionNumber int32,
```
Comment (naturetime10) [inline core/dao/app_package_upload_session.go]:
```suggestion
		versionNumber int,
```

File: core/dao/app_package_upload_session.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+		fileUploadSessionID uint64,
+	) (entity.AppPackageUploadSession, *errs.Error)
+	CreateAppPackageUploadSession(ct context.Context, tx *transaction.Transaction, AppPackageUploadSession entity.AppPackageUploadSession) *errs.Error
```
Comment (naturetime10) [inline core/dao/app_package_upload_session.go]:
```suggestion
	CreateAppPackageUploadSession(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    session entity.AppPackageUploadSession,
	) *errs.Error
```

File: core/dao/app_package_upload_session.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	UpdateAppPackageFileUploadSession(ct context.Context, tx *transaction.Transaction, AppPackageUploadSession entity.AppPackageUploadSession) *errs.Error
```
Comment (naturetime10) [inline core/dao/app_package_upload_session.go]:
```suggestion
	UpdateAppPackageFileUploadSession(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    session entity.AppPackageUploadSession,
	) *errs.Error
```

File: core/dao/daotest/app_package_upload_session.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			return &errs.Error{
+				Code:    errs.NotFound,
+				Message: fmt.Sprint("row not found: appID=%v, userID=%v, versionNumber=%v, fileUploadSessionID=%v", appID, userID, versionNumber, fileUploadSessionID),
+			}
+		},
+	})
+	return uploadSession, err
+}
+
+func (a *AppPackageUploadSession) CreateAppPackageUploadSession(ct context.Context, tx *transaction.Transaction, appPackageUploadSession entity.AppPackageUploadSession) *errs.Error {
```
Comment (naturetime10) [inline core/dao/daotest/app_package_upload_session.go]:
```suggestion
func (a *AppPackageUploadSession) CreateAppPackageUploadSession(ct context.Context, tx *transaction.Transaction, session entity.AppPackageUploadSession) *errs.Error {
```

File: core/dao/daotest/app_package_upload_session.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+					}
+				}
+			}
+
+			table.Rows = append(table.Rows, appPackageUploadSession)
+			return nil
+		},
+		Undo: func() *errs.Error {
+			table, err := a.db.GetTable(AppPackageUploadSessionTableName)
+
```
Comment (naturetime10) [inline core/dao/daotest/app_package_upload_session.go]:
```suggestion
```

File: core/dao/daotest/app_package_upload_session.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+				rows = append(rows, rawRow)
+			}
+
+			table.Rows = rows
+			return nil
+		},
+	})
+}
+
+func (a *AppPackageUploadSession) UpdateAppPackageFileUploadSession(ct context.Context, tx *transaction.Transaction, appPackageUploadSession entity.AppPackageUploadSession) *errs.Error {
```
Comment (naturetime10) [inline core/dao/daotest/app_package_upload_session.go]:
```suggestion
func (a *AppPackageUploadSession) UpdateAppPackageFileUploadSession(
    ct context.Context, 
    tx *transaction.Transaction, 
    session entity.AppPackageUploadSession,
) *errs.Error {
```

File: core/dao/daotest/app_package_upload_session.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	return tx.ExecuteCommand(transaction.Command{
+		Execute: func() *errs.Error {
+			table, err := a.db.GetTable(AppPackageUploadSessionTableName)
+			if err != nil {
+				return err
+			}
+
+			for i, rawRow := range table.Rows {
```
Comment (naturetime10) [inline core/dao/daotest/app_package_upload_session.go]:
```suggestion
			for index, rawRow := range table.Rows {
```

File: core/dao/daotest/app_package_upload_session.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+					appPackageUploadSession.VersionNumber),
+			}
+		},
+		Undo: func() *errs.Error {
+			table, err := a.db.GetTable(AppPackageUploadSessionTableName)
+			if err != nil {
+				return err
+			}
+
+			for i, rawRow := range table.Rows {
```
Comment (naturetime10) [inline core/dao/daotest/app_package_upload_session.go]:
```suggestion
			for index, rawRow := range table.Rows {
```

File: core/dao/sqldb/app_package_upload_session.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		SELECT
+			app_id,
+			user_id,
+			file_upload_session_id,
+			version_number,
+			is_completed,
+			created_at,
+			updated_at
+		FROM app_package_upload_session
+		WHERE app_id = $1 AND user_id = $2 AND version_number = $3 and file_upload_session_id = $4;`,
```
Comment (naturetime10) [inline core/dao/sqldb/app_package_upload_session.go]:
```suggestion
		WHERE app_id = $1 
		    AND user_id = $2 
		    AND version_number = $3 
		    AND file_upload_session_id = $4;`,
```

File: core/dao/sqldb/app_package_upload_session.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		Scan(
+			&appPackageUploadSession.AppID,
+			&appPackageUploadSession.UserID,
+			&appPackageUploadSession.FileUploadSessionID,
+			&appPackageUploadSession.VersionNumber,
+			&appPackageUploadSession.IsCompleted,
+			&appPackageUploadSession.CreatedAt,
+			&appPackageUploadSession.UpdatedAt,
+		)
+
```
Comment (naturetime10) [inline core/dao/sqldb/app_package_upload_session.go]:
```suggestion
```

File: core/dao/sqldb/app_package_upload_session.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	if errors.Is(err, sql.ErrNoRows) {
+		return entity.AppPackageUploadSession{}, errs.NewError(errs.NotFound, fmt.Sprintf("AppPackageUploadSession not found: appID=%v, userID=%v, versionNumber=%v",
```
Comment (naturetime10) [inline core/dao/sqldb/app_package_upload_session.go]:
```suggestion
		return entity.AppPackageUploadSession{}, errs.NewError(errs.NotFound, fmt.Sprintf("AppPackageUploadSession not found: appID=%v, userID=%v, versionNumber=%v",
```

File: core/dao/sqldb/app_package_upload_session.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		return errs.NewError(errs.Unknown, err.Error())
+	}
+
+	return nil
+}
+
+func (a *AppPackageUploadSession) UpdateAppPackageFileUploadSession(
+	ct context.Context,
+	tx *transaction.Transaction,
+	AppPackageUploadSession entity.AppPackageUploadSession) *errs.Error {
```
Comment (naturetime10) [inline core/dao/sqldb/app_package_upload_session.go]:
```suggestion
	session entity.AppPackageUploadSession,
) *errs.Error {
```

File: migrations/20230816221458_app_package_upload_session.sql [hunk 13]
```diff
@@ -0,0 +1,13 @@
+-- +migrate Up
+CREATE TABLE app_package_upload_session (
+	app_id BIGINT NOT NULL REFERENCES app (id),
+    user_id BIGINT NOT NULL REFERENCES "user" (id),
+    file_upload_session_id BIGINT NOT NULL,
+	version_number INT NOT NULL,
+	is_completed BOOLEAN NOT NULL DEFAULT FALSE,
+	created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	updated_at TIMESTAMP
```
Comment (naturetime10) [inline migrations/20230816221458_app_package_upload_session.sql]:
```suggestion
	app_id BIGINT NOT NULL REFERENCES app (id),
        user_id BIGINT NOT NULL REFERENCES "user" (id),
        file_upload_session_id BIGINT NOT NULL,
	version_number INT NOT NULL,
	is_completed BOOLEAN NOT NULL DEFAULT FALSE,
	created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
	updated_at TIMESTAMP
```

Comment (naturetime10) [reply migrations/20230816221458_app_package_upload_session.sql]: is `user_id` needed?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: migrations/20230816221458_app_package_upload_session.sql (hunk 13 above)
Comment (duanxianbo) [reply migrations/20230816221458_app_package_upload_session.sql]: no, will remove it

**Comment by naturetime10** [comment]
There is merge conflicts in this PR.

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**